### PR TITLE
[FIX] Add qMRI fmap filename rules and reorganize fmap patterns

### PR DESCRIPTION
**Pull request by tsalo**
Stems from https://github.com/bids-standard/bids-examples/pull/245#issuecomment-776167347.

Changes proposed:
- Add qMRI field map patterns (TB1EPI and TB1DAM).
- Reorganize field map patterns based on schema. This will make the `dir` entity required for `_epi` and `_m0scan` field maps.

### CODE REVIEW

**Comment by tsalo** [comment]
The part I'm stuck on is how the `fieldmapMainNiiData` comes into play. Should I just recreate the general field map patterns without the `json` suffix?

**Comment by effigies** [comment]
`isFieldMapMainNii` was added in #393 to check whether a file should/may have `IntendedFor` metadata.

**Comment by tsalo** [comment]
Thanks! It sounds like replicating the full set of patterns, minus jsons, was the way to go then.

**Comment by effigies** [comment]
What if we just say `isFieldMap(...) and not path[-5:] == ".json"`? (Sorry for dropping into Python...)

**Comment by tsalo** [comment]
I know so little about JavaScript that I'm not sure how to do that, but the general idea sounds solid to me.

**Comment by effigies** [comment]
[`endsWith()`](https://www.w3schools.com/Jsref/jsref_endswith.asp)

Will make a code recommendation, and we can see if it works...

File: bids-validator/utils/type.js [hunk 1]
```diff
@@ -206,11 +212,21 @@ export default {
...
     },
 
     isFieldMapMainNii: function(path) {
-      return conditionalMatch(fieldmapMainNiiData, path)
+      return (
+        conditionalMatch(fmapGreNii, path) ||
+        conditionalMatch(fmapPepolarAslNii, path) ||
+        conditionalMatch(fmapTB1DAMNii, path) ||
+        conditionalMatch(fmapTB1EPINii, path)
+      )
```
Comment (effigies) [inline bids-validator/utils/type.js]:
```suggestion
      return (
        !(path.endsWith('.json')) && (
          /* isFieldMap */
          conditionalMatch(fmapGre, path) ||
          conditionalMatch(fmapPepolarAsl, path) ||
          conditionalMatch(fmapTB1DAM, path) ||
          conditionalMatch(fmapTB1EPI, path)
          )
      )
```

File: bids-validator/utils/type.js [hunk 2]
```diff
@@ -32,8 +32,14 @@ const anatMTR = buildRegExp(file_level_rules.anat_mtr)
...
-const fieldmapData = buildRegExp(file_level_rules.field_map)
-const fieldmapMainNiiData = buildRegExp(file_level_rules.field_map_main_nii)
+const fmapGre = buildRegExp(file_level_rules.fmap_gre)
+const fmapPepolarAsl = buildRegExp(file_level_rules.fmap_pepolar_asl)
+const fmapTB1DAM = buildRegExp(file_level_rules.fmap_TB1DAM)
+const fmapTB1EPI = buildRegExp(file_level_rules.fmap_TB1EPI)
+const fmapGreNii = buildRegExp(file_level_rules.fmap_gre_nii)
+const fmapPepolarAslNii = buildRegExp(file_level_rules.fmap_pepolar_asl_nii)
+const fmapTB1DAMNii = buildRegExp(file_level_rules.fmap_TB1DAM_nii)
+const fmapTB1EPINii = buildRegExp(file_level_rules.fmap_TB1EPI_nii)
```
Comment (effigies) [inline bids-validator/utils/type.js]:
```suggestion
```

File: bids-validator/utils/type.js [hunk 3]
```diff
@@ -206,11 +212,24 @@ export default {
...
-      return conditionalMatch(fieldmapMainNiiData, path)
+      return (
+        !(path.endsWith('.json')) && (
+          /* isFieldMap */
+          conditionalMatch(fmapGre, path) ||
+          conditionalMatch(fmapPepolarAsl, path) ||
+          conditionalMatch(fmapTB1DAM, path) ||
+          conditionalMatch(fmapTB1EPI, path)
+          )
+      )
```
Comment (effigies) [inline bids-validator/utils/type.js]:
```suggestion
        !path.endsWith('.json') &&
        (
          /* isFieldMap */
          conditionalMatch(fmapGre, path) ||
          conditionalMatch(fmapPepolarAsl, path) ||
          conditionalMatch(fmapTB1DAM, path) ||
          conditionalMatch(fmapTB1EPI, path))
      )
```

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 4]
```diff
@@ -208,25 +208,87 @@
...
+    }
+  },
+
+  "fmap_gre_nii": {
+    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_run-[0-9]+)?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
     "tokens": {
       "@@@_field_map_type_@@@": [
         "phasediff",
         "phase1",
         "phase2",
```
Comment (effigies) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
```

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 5]
```diff
@@ -208,25 +208,87 @@
...
+    }
+  },
+
+  "fmap_pepolar_asl_nii": {
+    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_ce-[a-zA-Z0-9]+)?_dir-[a-zA-Z0-9]+(?:_run-[0-9]+)?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
+    "tokens": {
+      "@@@_field_map_type_@@@": [
+        "m0scan",
         "epi"
       ],
```
Comment (effigies) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
```

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 6]
```diff
@@ -208,25 +208,87 @@
...
+    }
+  },
+
+  "fmap_TB1EPI_nii": {
+    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_ce-[a-zA-Z0-9]+)?(?:_rec-[a-zA-Z0-9]+)(?:_run-[0-9]+)?_echo-[0-9]+?_flip-[0-9]+?(?:_part-(imag|mag|phase|real))?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
+    "tokens": {
+      "@@@_field_map_type_@@@": [
+        "TB1EPI"
+      ],
+      "@@@_field_map_ext_@@@": ["nii\\.gz", "nii", "json"]
```
Comment (effigies) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
```

**Review by effigies — COMMENTED** [review]
I think this should fix the linting errors...

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 7]
```diff
@@ -208,25 +208,43 @@
...
+    }
+  },
+
     }
   },
```
Comment (effigies) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
```

File: bids-validator/utils/type.js [hunk 8]
```diff
@@ -206,11 +208,24 @@ export default {
...
     },
 
     isFieldMapMainNii: function(path) {
-      return conditionalMatch(fieldmapMainNiiData, path)
+      return (
+        !path.endsWith('.json') &&
+        (
+          /* isFieldMap */
+          conditionalMatch(fmapGre, path) ||
```
Comment (effigies) [inline bids-validator/utils/type.js]:
```suggestion
        /* isFieldMap */
        (conditionalMatch(fmapGre, path) ||
```

**Review by effigies — APPROVED** [review]

**Pull request merged**